src/functTest/groovy/com/bmuschko/gradle/docker/AbstractFunctionalTest.groovy

```diff
@@ -63,6 +88,105 @@ abstract class AbstractFunctionalTest extends Specification {
         result.output.contains('Retrieving Docker version.')
     }
 
+    def setupSpec() {
+        dockerImagesBeforeTests = dockerClient.listImagesCmd().exec() as Set
+        dockerContainersBeforeTests = dockerClient.listContainersCmd().withShowAll(true).exec() as Set
+    }
+
+    def cleanupSpec() {
+        removeDockerContainers()
+        removeDockerImages()
+    }
+
+    private void removeDockerContainers() {
+
+        def dockerContainersAfterTests = dockerClient.listContainersCmd().withShowAll(true).exec() as Set
+
+        def removableDockerContainerIds = dockerContainersAfterTests.collect { it.id } - dockerContainersBeforeTests.collect { it.id }
+
+        println("Removing ${removableDockerContainerIds.size()} containers: ${removableDockerContainerIds}")
+        removableDockerContainerIds.each {
+            try {
+                dockerClient.removeContainerCmd(it)
+                    .withForce(true)
+                    .exec()
+            } catch (NotFoundException ex) {
+                println("Exception while removing container: ${ex.message} ")
+            }
+        }
+
+        println("Docker containers removed.")
+    }
+
+    private void removeDockerImages() {
+
+        def dockerImagesCreatedAfterTests = dockerClient.listImagesCmd().exec() as Set
+
+        def removableDockerImageIds = dockerImagesCreatedAfterTests.collect { it.id }  - dockerImagesBeforeTests.collect { it.id }
+
+        println("Removing ${removableDockerImageIds.size()} images: ${removableDockerImageIds}")
+
+        removableDockerImageIds.each {
+            try {
+                dockerClient.removeImageCmd(it)
+                    .withForce(true)
+                    .exec()
+            } catch (NotFoundException ex) {
+                println("Exception while removing image: ${ex.message} ")
+            }
+        }
+
+        println("Docker images removed.")
+    }
+
+    protected void startDockerRegistryContainer() {
+
+        def containers = dockerClient.listContainersCmd().exec()
+        println("Active dockerContainersBeforeTests: ${containers}")
+
+        boolean isRegistryAlreadyRunning
+
+        isRegistryAlreadyRunning = containers.find {
+            return it?.names?.contains("/registry")
```

Comment:
in most cases this should work, however if I start registry container with custom name it will fail. But for now it's :ok_hand: